pycqed/instrument_drivers/physical_instruments/QuTech_QCC.py

```diff
@@ -0,0 +1,615 @@
+        self.stop()
+        if not isinstance(filename, str):
+            raise ValueError(
+                "The parameter filename type({}) is incorrect. "
+                "It should be str.".format(type(filename)))
+
+        success_parser = self.QISA.assemble(filename)
+
+        if success_parser is not True:
+            print("Error detected while assembling the file {}:".format(
+                filename))
+            print(self.QISA.getLastErrorMessage())
+            raise RuntimeError("Assembling failed.")
+
+        instHex = self.QISA.getInstructionsAsHexStrings(False)
+
+        intarray = []
+        for instr in instHex:
+            intarray.append(int(instr[2:], 16))
+
+        # add a stop instruction at the end of the program
+        intarray.append(0x10000000)
+
+        if len(intarray) > MAX_NUM_INSN:
+            raise OverflowError("Failed to upload instructions: program length ({})"
+                " exceeds allowed maximum value ({}).".format(len(intarray),
+                    MAX_NUM_INSN))
+            return
+
+
+        binBlock = bytearray(array.array('L', intarray))
+        # print("binblock size:", len(binBlock))
+        # write binblock
+        hdr = 'QUTech:UploadInstructions '
+        self.binBlockWrite(binBlock, hdr)
+        # print("QCC: Sending instructions to the hardware finished.")
+
+        # write to last_loaded_instructions so it can conveniently be read back
+        self.last_loaded_instructions(filename)
+
+    def _upload_microcode(self, filename):
+        """
+        _upload_controls is different from send_instructions because we can
+        generate the microcode from a text file and the generation of the
+        microcode is done by the QCCMicrocode.py
+        """
+
+        if not isinstance(filename, str):
+            raise ValueError(
+                "The parameter filename type({}) is incorrect. "
+                "It should be str.".format(type(filename)))
+
+        self.microcode.load_microcode(filename)
+        binBlock = self.microcode.write_to_bin()
+        if not isinstance(binBlock, bytearray):
+            raise ValueError(
+                "The parameter binBlock type({}) is incorrect. "
+                "It should be bytearray.".format(type(binBlock)))
+
+        # write binblock
+        hdr = 'QUTech:UploadMicrocode '
+        self.binBlockWrite(binBlock, hdr)
+
+    def _upload_opcode_qmap(self, filename: str):
+        success = self.QISA.loadQuantumInstructions(filename)
+        if not success:
+            logging.warning("Error: ", self.QISA.getLastErrorMessage())
+            logging.warning("Failed to load quantum instructions from dictionaries.")
+
+        return success
+
+    def _set_vsm_chan_delay(self, chanNum, value):
+        """
+        This function is available for the user to 'hack' the
+        vsm_channel_delay using just a single function name
+        """
+        self.write('QUTech:VSMChannelDelay%d %d' % (chanNum, value))
```

Comment:
Maybe add some kind of validator to `value` and `chanNum`. What happens if the user sends an invalid value?

Reply:
Validation is provided at device driver level. Therefore, maintaining python driver as is for now